### PR TITLE
Doc: PR Guidance on WIP, Draft and LLMs.

### PR DESCRIPTION
**Pull request by ax3l**
Improve our PR guidance on workflows for WIP, Draft and self-review cycles with LLMs.

### CODE REVIEW

File: .github/pull_request_template.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+</details>
+
+## TODO
+
+<!--
+Prefix the PR title with `[WIP] ` (and/or open the PR as a GitHub *draft*)
+until this list is complete. CI is paused on draft PRs.
+
+If this PR was drafted with an LLM, perform a separate LLM-based review on to of your manual self-review before asking for a human reviewer to read your code.
```
Comment (lucafedeli88) [inline .github/pull_request_template.md]: "drafted with an LLM" may be a bit confusing, as it puts in the same category the scenarios "drafted mainly by an LLM" and "drafted with minimal LLM assistance". We may want to set the bar somewhere in this continuum (eventually at one of the extremes).

Comment (lucafedeli88) [inline .github/pull_request_template.md]:
```suggestion
If this PR was drafted with an LLM, perform a separate LLM-based review on top of your manual self-review before asking for a human reviewer to read your code.
```

small typo

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+-->
+
+- [ ] Self-reviewed: check and understand logic, improve wording, check for outdated and unnecessarily historic documentation; ensure new features are tested with small, fast tests
+- [ ] Finalize: ensure a WarpX developer can understand this
```
Comment (lucafedeli88) [inline .github/pull_request_template.md]: "ensure a WarpX developer can understand this" maybe could be rephrased as something like
"strive to make the PR (code and description) understandable for WarpX developers"

File: AGENTS.md [hunk 3]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
 - Main branch: `development` (not `main`)
 - Fork-and-branch workflow; PRs target `development`
 - Pull requests with new features need to add a test for coverage.
+
+## Pull Requests
```
Comment (lucafedeli88) [inline AGENTS.md:151]: Maybe we could we refer to `pull_request_template.md` here instead of duplicating its content.
We could keep here only the additional parts required for agents.

**Review by lucafedeli88 — CHANGES_REQUESTED** [review]
Thanks, @ax3l !
I've left few small comments.

File: .github/pull_request_template.md (hunk 1 above)
Comment (jlvay) [reply .github/pull_request_template.md]: I agree. Alternatively, one could say, e.g., "If any portion of code (even one character) has been drafted by an LLML and not reviewed by a human, ...".

File: AGENTS.md [hunk 4]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+
+Close the description with a `## TODO` check-list and keep the PR title prefixed with `[WIP] ` until the list is complete:
+
+```markdown
+## TODO
+
+Prefix the PR title with `[WIP] ` until this list is complete.
+
+- [x] Drafted with LLM [optional model details]
+- [ ] Manual self-review: check and understand logic, improve wording, check for outdated and unnecessarily historic documentation; ensure new features are tested with small, fast tests
```
Comment (jlvay) [inline AGENTS.md]:
```suggestion
- [ ] Manual self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features were tested with small, fast tests
```

File: AGENTS.md [hunk 5]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+- [ ] Performed a separate, LLM-based review and fixed all findings
+- [ ] Finalize: present and documented in a way to ensure a human WarpX developer can understand this
```
Comment (jlvay) [inline AGENTS.md]:
```suggestion
- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand 
```

File: AGENTS.md [hunk 6]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+```
+
+WarpX developers are typically a mix of backgrounds from computational physics, applied mathematics, and computer systems engineering.
```
Comment (jlvay) [inline AGENTS.md]:
```suggestion
WarpX developers have typically a mix of backgrounds from computational physics, applied mathematics, and computer systems engineering.
```

File: .github/pull_request_template.md (hunk 1 above)
Comment (ax3l) [reply .github/pull_request_template.md]: Good points. I would suggest:
```suggestion
If any portion of code has been drafted by an LLM, perform a separate LLM-based review on top of your manual self-review before asking for a human reviewer to read your code.
```

I find the "reviewed by a human" contradicts a bit our ask to also make a human self-review, and is a not wordy :)

File: .github/pull_request_template.md (hunk 2 above)
Comment (ax3l) [inline .github/pull_request_template.md]:
```suggestion
- [ ] Self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features were tested with small, fast tests
- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand
```

File: AGENTS.md (hunk 3 above)
Comment (ax3l) [reply AGENTS.md:151]: The PR template does intentionally skip the LLM parts, because not all PRs (yet) are LLM authored.

This might change in the future, but some developers raised this on Tuesday as not to assume by default for now.

File: .github/pull_request_template.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+## TODO
+
+<!--
+Prefix the PR title with `[WIP] ` (and/or open the PR as a GitHub *draft*)
+until this list is complete. CI is paused on draft PRs.
+
+If any portion of code has been drafted by an LLM, perform a separate LLM-based review on top of your manual self-review before asking for a human reviewer to read your code.
+-->
+
+- [ ] Self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features were tested with small, fast tests
```
Comment (EZoni) [inline .github/pull_request_template.md]: I think the present tense here signals better that tests should be included and automated, rather than being a one-time test performed once manually during the period in which the author worked on the PR:
```suggestion
- [ ] Self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
```

File: .github/pull_request_template.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand
```
Comment (EZoni) [inline .github/pull_request_template.md]: I feel like this could also be folded into the self-review step above, but I don't have strong opinions.

File: AGENTS.md (hunk 3 above)
Comment (EZoni) [reply AGENTS.md:151]: On the other hand, it depends on what "LLM authored" means. LLMs at this point are injecting suggestions even through auto-completion/auto-suggestion in most basic IDE workflows, so unless a developer uses none of these tools, I think one could claim that it's reasonable to expect some LLM footprint nowadays, even if it wasn't a fully agentic implementation.

File: AGENTS.md [hunk 9]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+
+Close the description with a `## TODO` check-list and keep the PR title prefixed with `[WIP] ` until the list is complete:
+
+```markdown
+## TODO
+
+Prefix the PR title with `[WIP] ` until this list is complete.
+
+- [x] Drafted with LLM [optional model details]
+- [ ] Manual self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features were tested with small, fast tests
```
Comment (EZoni) [inline AGENTS.md]:
```suggestion
- [ ] Manually self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
```

File: .github/pull_request_template.md [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+corner-cases and rejected ideas into the collapsed "Details" block below.
+-->
+
+## What was achieved?
+
+
+## What was tested?
+
+
+## Does this change break user interfaces?
```
Comment (EZoni) [inline .github/pull_request_template.md]: Maybe simpler while still clear, but no strong opinions:

```suggestion
## Does this break user interfaces?
```

File: AGENTS.md [hunk 11]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
 - Pull requests with new features need to add a test for coverage.
+
+## Pull Requests
+
+Focus the PR description on the new feature and the tests performed.
+Stay below one page on GitHub, i.e., less than 15 lines, and answer:
+
+- What was achieved?
+- What was tested?
+- Does this change break user interfaces?
```
Comment (EZoni) [inline AGENTS.md]: Maybe simpler while still clear, but no strong opinions:

```suggestion
- Does this break user interfaces?
```

File: CONTRIBUTING.rst [hunk 12]
```diff
@@ -174,12 +174,26 @@ Right after you push changes, a banner should appear on the Github page of your
...
-  People who review your PR are happy to know
+  The description is pre-filled with our `pull request template <https://github.com/BLAST-WarpX/warpx/blob/development/.github/pull_request_template.md>`__, which asks the three questions that reviewers need answered:
 
-  * what feature/fix you propose, and why
-  * how you made it (added new/edited files, created a new class than inherits from...)
-  * how you tested it and what was the output you got
-  * and anything else relevant to your PR (attach images and scripts, link papers, *etc.*)
+  * **What was achieved?** The feature/fix you propose, and why.
+  * **What was tested?** How you tested it and what output you got.
+  * **Does this change break user interfaces?** For example: renamed or removed input parameters, changed defaults, or changed Python APIs.
```
Comment (EZoni) [inline CONTRIBUTING.rst]: Maybe simpler while still clear, but no strong opinions:

```suggestion
  * **Does this break user interfaces?** For example: renamed or removed input parameters, changed defaults, or changed Python APIs.
```

File: AGENTS.md (hunk 3 above)
Comment (lucafedeli88) [reply AGENTS.md:151]: > The PR template does intentionally skip the LLM parts, because not all PRs (yet) are LLM authored.
> 
> This might change in the future, but some developers raised this on Tuesday as not to assume by default for now.

Sorry, my comment was not very clear. I fully agree that we should not assume that all PR are LLM authored (as far as I am concerned, I am not using them at all). What I mean is that there's some text duplication between `pull_request_template.md` and `AGENTS.md` . I was wondering if it could be practical to have a PR template for everyone (humans and agents) in `pull_request_template.md` plus an additional part exclusively for agents (in  `AGENTS.md`). If it's not practical, I am OK with duplication.

**Comment by EZoni** [comment]
Another thing that came to my mind and was already discussed in the past: I think we should add some language about rebasing with care (as opposed to merging), to make reviewers' life easier, especially given that rebasing usually invalidates the "Changes since your last review" feature under "Select commits to view".

File: .github/pull_request_template.md [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+</details>
+
+## TODO
+
+<!--
+Prefix the PR title with `[WIP] ` (and/or open the PR as a GitHub *draft*)
+until this list is complete. CI is paused on draft PRs.
+
+If any portion of code has been drafted by an LLM, perform a separate LLM-based review on top of your manual self-review before asking for a human reviewer to read your code.
```
Comment (EZoni) [inline .github/pull_request_template.md]: Actually, why is this just in a comment (invisible when rendered) rather than regular text?

File: AGENTS.md [hunk 14]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+Close the description with a `## TODO` check-list and keep the PR title prefixed with `[WIP] ` until the list is complete:
+
+```markdown
+## TODO
+
+Prefix the PR title with `[WIP] ` until this list is complete.
+
+- [x] Drafted with LLM [optional model details]
+- [ ] Manually self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
+- [ ] Performed a separate, LLM-based review and fixed all findings
```
Comment (RemiLehe) [inline AGENTS.md]: "fixed all findings" is too amibitious, no? LLM-based review will always find some small nit-picking things in my experience.

File: .github/pull_request_template.md (hunk 13 above)
Comment (RemiLehe) [inline .github/pull_request_template.md]: This text puts the emphasis on LLM-based review, and de-emphasize the human review ("on top of..."). In addition, "manual self-review" can be somewhat ambiguous, and some contributors might necessarily understand that this means that they need to read the code.

I would instead be very direct about the human review:
"If any portion of code has been drafted by an LLM, **the human that controls this LLM should read and review the code, and make sure it is understandable** before asking another human to review."

Comment (ax3l) [reply .github/pull_request_template.md]: Because we did not want to assume, per earlier review, that every PR is LLM-assisted. This was the compromise to not have it in the check-list.

Comment (ax3l) [reply .github/pull_request_template.md]: Yes, that is what the first bullet on self-review in the shown todo list (L33) is trying to express (we can rephrase that one)

File: AGENTS.md (hunk 3 above)
Comment (ax3l) [reply AGENTS.md:151]: Yep, I don't like the duplication... I guess we could embed the PR template here via reference, but then we still need to highlight the difference for the AGENT.md

File: AGENTS.md (hunk 14 above)
Comment (ax3l) [reply AGENTS.md]: I just let it iterate until it does not introduce new bugs and docs are not stale :D

How about?

```suggestion
- [ ] Performed a separate, LLM-based review loop: no new bugs and new stale docs remain
```

File: .github/pull_request_template.md [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+-->
+
+- [ ] Self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
+- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand
```
Comment (ax3l) [inline .github/pull_request_template.md]:
```suggestion
- [ ] **Self-reviewed:** I read and understand the new code and documentation
```

File: AGENTS.md [hunk 16]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
 - Pull requests with new features need to add a test for coverage.
+
+## Pull Requests
+
+Focus the PR description on the new feature and the tests performed.
+Stay below one page on GitHub, i.e., less than 15 lines, and answer:
+
+- What was achieved?
+- What was tested?
+- Does this break user interfaces?
```
Comment (ax3l) [inline AGENTS.md]:
```suggestion
Make sure the PR description is concise.
```

File: AGENTS.md [hunk 17]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+
+Put additional system details, corner-cases and rejected ideas in a collapsed block, so the summary stays short:
```
Comment (ax3l) [inline AGENTS.md]:
```suggestion
If important, add system details, corner-cases and rejected ideas in a collapsed block, so the summary stays short:
```

File: AGENTS.md [hunk 18]
```diff
@@ -147,3 +147,38 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+
+```markdown
+<details><summary>Details</summary>
```
Comment (ax3l) [inline AGENTS.md]:
```suggestion
<details>
```

File: AGENTS.md [hunk 19]
```diff
@@ -147,3 +147,33 @@ When a change removes or renames a user-facing input parameter, add a guard to t
 - Main branch: `development` (not `main`)
 - Fork-and-branch workflow; PRs target `development`
 - Pull requests with new features need to add a test for coverage.
+
+## Pull Requests
+
+Make sure the PR description is concise.
```
Comment (ax3l) [inline AGENTS.md:153]:
```suggestion
If you have not done so already: performed a separate code review loop of the changes with sub-agents until no new bugs and new stale docs remain.
New inline and API doc strings must not have reflective or historic code state documented.
Only after that, open a PR.

Make sure the PR description is concise.
```

File: AGENTS.md [hunk 20]
```diff
@@ -147,3 +147,33 @@ When a change removes or renames a user-facing input parameter, add a guard to t
...
+```markdown
+## TODO
+
+Prefix the PR title with `[WIP] ` until this list is complete.
+
+- [x] Drafted with LLM [optional model details]
+- [ ] Manually self-reviewed: checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
+- [ ] Performed a separate, LLM-based review loop: no new bugs and new stale docs remain
+- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand
+```
```
Comment (ax3l) [inline AGENTS.md]:
```suggestion
Prefix the PR title with `[WIP] ` until this list is complete.

- [x] Drafted with LLM [optional model details]
- [ ] **Self-reviewed**: I checked and understand logic, improved wording, checked for outdated and unnecessarily historic documentation; ensured new features are tested with small, fast tests
- [ ] Performed a separate, LLM-based review loop: no new bugs and new stale docs remain
- [ ] Finalized: documented in a way to ensure a human WarpX developer can understand
```
```

File: AGENTS.md [hunk 21]
```diff
@@ -147,3 +147,32 @@ When a change removes or renames a user-facing input parameter, add a guard to t
 - Main branch: `development` (not `main`)
 - Fork-and-branch workflow; PRs target `development`
 - Pull requests with new features need to add a test for coverage.
+
+## Pull Requests
```
Comment (roelof-groenewald) [inline AGENTS.md:151]: Should we add a point here to ask the LLM to check if the PR can be sensibly split into multiple smaller ones? Alternatively such a point could go into #7155 as part of the self-review.
